chalk-solve/src/recursive/mod.rs

```diff
@@ -0,0 +1,520 @@
+                            },
+                        },
+                    };
+
+                    // "Domain" goals (i.e., leaf goals that are Rust-specific) are
+                    // always solved via some form of implication. We can either
+                    // apply assumptions from our environment (i.e. where clauses),
+                    // or from the lowered program, which includes fallback
+                    // clauses. We try each approach in turn:
+
+                    let InEnvironment { environment, goal } = &canonical_goal.canonical.value;
+
+                    let (prog_solution, prog_prio) = {
+                        debug_heading!("prog_clauses");
+
+                        let prog_clauses = self.program_clauses_for_goal(environment, &goal);
+                        match prog_clauses {
+                            Ok(clauses) => {
+                                self.solve_from_clauses(&canonical_goal, clauses, minimums)
+                            }
+                            Err(Floundered) => {
+                                (Ok(Solution::Ambig(Guidance::Unknown)), ClausePriority::High)
+                            }
+                        }
+                    };
+                    debug!("prog_solution={:?}", prog_solution);
+
+                    (prog_solution, prog_prio)
+                }
+
+                _ => {
+                    let canonical_goal = UCanonical {
+                        universes,
+                        canonical: Canonical {
+                            binders,
+                            value: InEnvironment { environment, goal },
+                        },
+                    };
+
+                    self.solve_via_simplification(&canonical_goal, minimums)
+                }
+            };
+
+            debug!(
+                "solve_new_subgoal: loop iteration result = {:?} with minimums {:?}",
+                current_answer, minimums
+            );
+
+            if !self.context.stack[depth].read_and_reset_cycle_flag() {
+                // None of our subgoals depended on us directly.
+                // We can return.
+                self.context.search_graph[dfn].solution = current_answer;
+                self.context.search_graph[dfn].solution_priority = current_prio;
+                return *minimums;
+            }
+
+            // Some of our subgoals depended on us. We need to re-run
+            // with the current answer.
+            if self.context.search_graph[dfn].solution == current_answer {
+                // Reached a fixed point.
+                return *minimums;
+            }
+
+            if (
+                self.context.search_graph[dfn].solution_priority,
+                current_prio,
+            ) == (ClausePriority::High, ClausePriority::Low)
+                && self.context.search_graph[dfn].solution.is_ok()
+            {
+                // TODO check solution inputs?
```

Comment:
So what is the role of this code, @flodiebold? I see there is also the `combine_with_priorities` code -- it seems like we want to use that here.

Also, total nit, but I find this hard to read. The `is_ok()` check is kind of obscured, and the tuple makes me jump back and forth to figure out what is compared to what. I'd prefer:

```rust
if (
    current_prio = ClausePriority::Low &&
    self.context.search_graph[dfn].solution_priority == ClausePriority::High &&
    self.context.search_graph[dfn].solution.is_ok()
) {
```

Reply:
I wasn't sure about using `combine_with_priorities` since the code generally doesn't combine solutions from earlier iterations here, it just overwrites them. But maybe that was equivalent so far :thinking:

Reply:
Using `combine_with_priorities` here is actually kind of complicated, because `combine_with_priorities` only works with a `DomainGoal` and `Solution`s, and here we have a `Goal` and `Result<Solution>`s, so we have to handle a bunch more cases :thinking:

Reply:
@nikomatsakis done, though as mentioned more complicated than it seemed at first...